src/pip/_internal/req/req_file.py

```diff
@@ -545,7 +546,23 @@ def get_file_content(url: str, session: "PipSession") -> Tuple[str, str]:
     # Assume this is a bare path.
     try:
         with open(url, "rb") as f:
-            content = auto_decode(f.read())
+            raw_content = f.read()
     except OSError as exc:
         raise InstallationError(f"Could not open requirements file: {exc}")
+
+    try:
+        content = auto_decode(raw_content)
+    except UnicodeDecodeError as exc:
+        fallback_encoding = "utf-8"
+        # don't try an decode again if we know it will fail
+        if exc.encoding == fallback_encoding:
+            raise
+
+        warnings.warn(
+            f"unable to decode data with {exc.encoding}, falling back to {fallback_encoding}",  #  noqa: E501
+            UnicodeWarning,
+            stacklevel=2,
+        )
```

Comment:
> The problem with using `warnings.warn` is that its presentation format is inappropriately technical. `logger.warning` should be used instead.

I think I just went with this because I knew `UnicodeWarning` was a thing, happy to go with logging [cae26c0](https://github.com/pypa/pip/pull/12795/commits/cae26c0c1f50a598e31cf715eb00bc79505188fb)